File: AI/BFS 8 puzzle/puzzle/src/BFS/BFS.java
package BFS;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Created by Rakshith on 2/2/2018.
 */
public class BFS {

    Queue<Node> frontier;
    List<Node> explored;

    public BFS(){
        frontier = new LinkedList<>();
        explored = new ArrayList<>();
    }

    public void bfs(Node root){

        Node currentlyPopped;

        frontier.add(root);
        root.setVisited(true);
        explored.add(root);

        while (!frontier.isEmpty()){
            currentlyPopped = frontier.poll();

            for (Node neighbour:currentlyPopped.getNeighbours()) {

                if(!neighbour.isVisited()){
                    frontier.add(neighbour);
                    neighbour.setVisited(true);
                    explored.add(neighbour);
                }

            }

        }

        System.out.println(explored);
    }

}

File: AI/BFS 8 puzzle/bfs python/Basicbfs/BFS.py
class BFS:
    def __init__(self):
        self.explored = []
        self.frontier = []

    def bfs(self, root, goal_value):
        self.frontier.append(root)
        root.set_visited(True)

        while len(self.frontier) != 0:
            popped_node = self.frontier.pop(0)
            self.explored.append(popped_node)

            if popped_node.get_data() == goal_value:
                return self.explored

            for neighbour in popped_node.get_neighbours():
                if (neighbour not in self.explored) and (not neighbour.is_visited()):
                    self.frontier.append(neighbour)
                    neighbour.set_visited(True)

        raise Exception("Couldn't find your goal.")
